src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/EnumerableConditionalWeakTable.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace Roslyn.Utilities;
+
+#if NET
+internal readonly struct EnumerableConditionalWeakTable<TKey, TValue>() : IEnumerable<KeyValuePair<TKey, TValue>>
+    where TKey : class
+    where TValue : class
+{
+    private readonly ConditionalWeakTable<TKey, TValue> _table = new();
+
+    public bool TryGetValue(TKey key, [NotNullWhen(true)] out TValue? value)
+        => _table.TryGetValue(key, out value);
+
+    public void Add(TKey key, TValue value)
+        => _table.Add(key, value);
+
+    public void AddOrUpdate(TKey key, TValue value)
+        => _table.AddOrUpdate(key, value);
+
+    public bool Remove(TKey key)
+        => _table.Remove(key);
+
+    public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
+        => ((IEnumerable<KeyValuePair<TKey, TValue>>)_table).GetEnumerator();
+
+    IEnumerator IEnumerable.GetEnumerator()
+        => ((IEnumerable)_table).GetEnumerator();
+}
+#else
+internal sealed class EnumerableConditionalWeakTable<TKey, TValue> : IEnumerable<KeyValuePair<TKey, TValue>>
+    where TKey : class
+    where TValue : class
+{
+    private sealed class Box(TKey key, TValue value)
```

Comment:
nit.  Consider just `StrongBox<KeyValuePair<>>`.  i personally think it's clearer since StrongBox is precisely for these scenarios.

Reply:
I did actually use it initially, but 1) it's mutable 2) the generic parameters makes the code noisy.